marvin: init at 19.1.0

fusion809 opened the pull request:
###### Motivation for this change
The Marvin Suite is one of the finest skeletal structure drawing programs that is available free of cost to the end user. There are some features one must pay for, but most of them are advanced features that many users, like myself, don't really care about. Here's a screenshot showing it running, opened to a molecule of clozapine:

![screenshot_20190112_220048](https://user-images.githubusercontent.com/4717341/51072949-93701a80-16b5-11e9-8841-1ea678fbe4da.png)

. I have added myself to the maintainers list, as this package, of course, needs a maintainer. 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Area that could do with some improvement
[Line 9 of the default.nix](https://github.com/fusion809/nixpkgs/blob/3f51913ca67eba6216d8458126cb2aafd0d0105c/pkgs/applications/science/chemistry/marvin/default.nix#L9) file could do with some improvement. Namely, if anyone knows a way I could let the nix file determine the version 19.1 from the full version 19.1.0, and use it in the URL, it'd be much appreciated :). I've tried using `${version%.?}` and sadly, it merely caused Nix to give the error message:

```bash
error: syntax error, unexpected $undefined, expecting '}', at /home/fusion809/GitHub/mine/packaging/nixpkgs/pkgs/applications/science/chemistry/marvin/default.nix:9:77
```

fusion809 commented:
Ticking the "other Linux distributions" box as I just tried it on Arch Linux and it is running fine.

symphorien commented on pkgs/applications/science/chemistry/marvin/default.nix:
```diff
@@ -0,0 +1,40 @@
+    mkdir $out/share/pixmaps
+    mkdir $out/bin
+    ln -sf $out/opt/chemaxon/marvinsuite/bin/msketch $out/bin/msketch
+    ln -sf $out/opt/chemaxon/marvinsuite/bin/mview $out/bin/mview
```
This does not work in an empty PATH. 
```
 PATH= $buildInputs/bin/mview 
/nix/store/jb299sb3abjqfh4zm7qpkd6ir0jl8jfc-marvin-19.1.0/bin/mview: line 362: basename: No such file or directory
/nix/store/jb299sb3abjqfh4zm7qpkd6ir0jl8jfc-marvin-19.1.0/bin/mview: line 363: dirname: No such file or directory
/nix/store/jb299sb3abjqfh4zm7qpkd6ir0jl8jfc-marvin-19.1.0/bin/mview: line 378: dirname: No such file or directory
/nix/store/jb299sb3abjqfh4zm7qpkd6ir0jl8jfc-marvin-19.1.0/bin/mview: line 379: basename: No such file or directory
No suitable Java Virtual Machine could be found on your system.
The version of the JVM must be at least 1.8.
Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.
```
It would be preferable to wrap these executables with the jre (and coreutils apparently). random example:
https://github.com/NixOS/nixpkgs/blob/b07e1563fe4cafd5e375a5861e42c10e0406b792/pkgs/applications/audio/abcde/default.nix#L38-L43
Or if running a jar is enough:
https://github.com/NixOS/nixpkgs/blob/b07e1563fe4cafd5e375a5861e42c10e0406b792/pkgs/applications/graphics/astah-community/default.nix#L52-L53

fusion809 replied:
The executing command, when the variables are expanded, is incredibly complex, e.g. here is what it is when I've extracted it to a local directory at /home/fusion809/Programs/Deb/marvin: https://gist.github.com/fusion809/188669fd39c100a2cfadc42c72774fb4.

fusion809 replied:
Oh thank goodness, I figured it out! Just had to find the docs for makeWrapper.

symphorien replied:
It still does not work for me in an empty PATH.
```
[nix-shell:~/.cache/nix-review/pr-53839]$ PATH= $buildInputs/bin/mview
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 362: basename: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 363: dirname: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 378: dirname: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 379: basename: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 66: expr: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 67: expr: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 81: awk: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 82: rm: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 83: mv: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 85: sed: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 88: chmod: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 66: expr: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 67: expr: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 81: awk: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 82: rm: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 83: mv: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 85: sed: No such file or directory
/nix/store/vs9h3hws3slbrliryxi80abmf48dws4j-marvin-19.1.0/opt/chemaxon/marvinsuite/bin/mview: line 88: chmod: No such file or directory
No suitable Java Virtual Machine could be found on your system.
The version of the JVM must be at least 1.8.
Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.
```
You probably need to add something like `--prefix PATH : ${makeBinPath [ coreutils awk sed ]}`.
And the INSTALL4J_JAVA_HOME thing does not seem to work.
By the way, I find strange that a startup wrapper needs rm...

fusion809 replied:
I added that to the wrapper commands and now building it gives:

```bash
rror: undefined variable 'makeBinPath' at /data/GitHub/mine/packaging/nixpkgs.marvin-pr/pkgs/applications/science/chemistry/marvin/default.nix:21:117
```

symphorien replied:
makeBinPath is in lib

fusion809 replied:
Replacing `makeBinPath` with `stdenv.lib.makeBinPath` makes the build fail, with the error:

```bash
error: undefined variable 'coreutils' at /data/GitHub/mine/packaging/nixpkgs.marvin-pr/pkgs/applications/science/chemistry/marvin/default.nix:21:142
```

Sorry, I'm sure these are real stupid concerns, but I'm a beginner and I do not know why this is happening.

symphorien replied:
coreutils needs to be added to the arguments in the first line.

fusion809 replied:
Thank you, I do believe I have fixed it. I tested with an empty PATH and it launched.

symphorien replied:
nice !

symphorien commented on pkgs/applications/science/chemistry/marvin/default.nix:
```diff
@@ -0,0 +1,40 @@
+  buildInputs = [ dpkg makeWrapper ];
+  unpackPhase = ''
+    mkdir pkg/opt -p
+    dpkg-deb -x $src pkg/opt
```
you can directly unpack to `$out/opt` if you wish so.

fusion809 replied:
Thanks, done in 1e14e08.

fusion809 commented:
In my last commit 355f8d8, I added extra executables, for all executables in `opt/chemaxon/marvinsuite`, so:

```bash
bin/cxcalc
bin/cxtrain
bin/evaluate
bin/molconvert
bin/msketch
bin/mview
LicenseManager
MarvinSketch
MarvinView
```

I also created a desktop config file for the ChemAxon License Manager.

symphorien commented on pkgs/applications/science/chemistry/marvin/default.nix:
```diff
@@ -0,0 +1,43 @@
+    makeWrapper $out/opt/chemaxon/marvinsuite/bin/evaluate $out/bin/mview --set INSTALL4J_JAVA_HOME "${jre}"
+    makeWrapper $out/opt/chemaxon/marvinsuite/bin/molconvert $out/bin/mview --set INSTALL4J_JAVA_HOME "${jre}"
+    makeWrapper $out/opt/chemaxon/marvinsuite/bin/msketch $out/bin/msketch --set INSTALL4J_JAVA_HOME "${jre}"
+    makeWrapper $out/opt/chemaxon/marvinsuite/bin/mview $out/bin/mview --set INSTALL4J_JAVA_HOME "${jre}"
```
Nit: You can probably write a loop like 
```
for i in cxcalc cxtrain etc; do
   makeWrapper $out/opt/chemaxon/marvinsuite/bin/$i $out/bin/$i --set INSTALL4J_JAVA_HOME "${jre}";
done
```

fusion809 replied:
Done in 1cdba3e048c.

fusion809 committed:
marvin: init at 19.1.0
Also adding myself (fusion809) as a maintainer.
The marvin Nix file in this commit is largely thanks to @msteen.
